LCLSGeneral/LCLSGeneral/POUs/Diagnostics/FB_EtherCATDiag.TcPOU

```diff
@@ -0,0 +1,575 @@
+
+	IF NOT fbGetAllSlaveStates.bError THEN
+		IF iNumOfSlavesRead = nSlaveCountCfg THEN
+			FOR I := 0 TO 3 DO
+				iDiagIndex[I] := 0;
+			END_FOR
+
+			(* split slave state and link state *)
+			FOR I := 0 TO MIN(iNumOfSlavesRead, iSLAVEADDR_ARR_SIZE) DO
+				(* slave state*)
+				arrSlaveInfo[I].stState.eEcState 		:= arrSlaveStates[I].deviceState AND 16#0F;
+				arrSlaveInfo[I].stState.bError 			:= arrSlaveStates[I].deviceState.4;
+				arrSlaveInfo[I].stState.bInvalidVPRS 	:= arrSlaveStates[I].deviceState.5;
+				(* link state *)
+				arrSlaveInfo[I].stState.bNoCommToSlave 	:= arrSlaveStates[I].linkState.0;
+				arrSlaveInfo[I].stState.bLinkError 		:= arrSlaveStates[I].linkState.1;
+				arrSlaveInfo[I].stState.bMissingLink 	:= arrSlaveStates[I].linkState.2;
+				arrSlaveInfo[I].stState.bUnexpectedLink := arrSlaveStates[I].linkState.3;
+				arrSlaveInfo[I].stState.bPortA 			:= arrSlaveStates[I].linkState.4;
+				arrSlaveInfo[I].stState.bPortB 			:= arrSlaveStates[I].linkState.5;
+				arrSlaveInfo[I].stState.bPortC 			:= arrSlaveStates[I].linkState.6;
+				arrSlaveInfo[I].stState.bPortD 			:= arrSlaveStates[I].linkState.7;
+				(* DiagData *)
+				arrSlaveInfo[I].bDiagData	:= ((arrSlaveStates[I].deviceState AND 16#F0) <> 0) OR
+					(((arrSlaveStates[I].deviceState AND 16#0F) > 0) AND ((arrSlaveStates[I].deviceState AND 16#0F) < 8)) OR
+					(arrSlaveStates[I].linkState <> 0);
+
+				IF arrSlaveInfo[I].bDiagData THEN
+					IF (I=0) THEN
+						iDiagIndex[0] := 0;
+					ELSE
+						IF (iDiagIndex[0] = 0) AND NOT arrSlaveInfo[0].bDiagData THEN
+							iDiagIndex[0] :=  UDINT_TO_UINT(I);
+						END_IF
+					END_IF
+				END_IF
+			END_FOR
+		END_IF
+	END_IF
+
+	IF arrSlaveInfo[iDiagIndex[0]].bDiagData THEN
+		iState := iState + 1;
+	ELSE
+		FOR I := 0 TO 3 DO
+			MEMSET(ADR(arrDiagSlaveInfo[I]), 0, SIZEOF(arrDiagSlaveInfo[I]));
+			MEMSET(ADR(arrDiagSlaveInfoScanned[I]), 0, SIZEOF(arrDiagSlaveInfoScanned[I]));
+		END_FOR
+		iState := 10;
+	END_IF
+END_IF
+]]></ST>
+      </Implementation>
+    </Action>
+    <Action Name="M_GetTopoData" Id="{e190db0a-cd90-44dc-b220-ccd10f73f62c}">
+      <Implementation>
+        <ST><![CDATA[fbGetTopologyData(
+	NETID	:= sMasterNetID,
+	PORT	:= 16#FFFF,
+	IDXGRP	:= 16#22,
+	IDXOFFS	:= 0,
+	LEN		:= iTopologyData*SIZEOF(arrTopologyData[0]),
+	DESTADDR:= ADR(arrTopologyData),
+	READ	:= TRUE,
+	TMOUT	:= tTimeout,
+);
+
+IF NOT fbGetTopologyData.BUSY THEN
+	fbGetTopologyData(READ := FALSE);
+
+	IF NOT fbGetTopologyData.ERR THEN
+		iDiagPort[0] := arrTopologyData[iDiagIndex[0]].iOwnPhysicalAddr;
+		iDiagPort[1] := arrTopologyData[iDiagIndex[0]].stPhysicalAddr.portB;
+		iDiagPort[2] := arrTopologyData[iDiagIndex[0]].stPhysicalAddr.portC;
+		iDiagPort[3] := arrTopologyData[iDiagIndex[0]].stPhysicalAddr.portD;
+
+		(* clear diag index  *)
+		iDiagIndex[1] := 0;
+		iDiagIndex[2] := 0;
+		iDiagIndex[3] := 0;
+
+		(* find slaves on PortB of first slave with diag *)
+		IF iDiagPort[1] <> 0 THEN
```

Comment:
3 nearly identical code blocks could be refactored into one. Loop over both `iDiagPort`, `arrToplogyData` and update `iDiagIndex`.